src/typedRequestData.utils.ts

```diff
@@ -0,0 +1,104 @@
+import {
+  MessageTypeProperty,
+  MessageTypes,
+  SignTypedDataVersion,
+  TypedDataUtils,
+  TypedMessage,
+} from '@metamask/eth-sig-util';
+import type { EnvelopingTypes } from '@rsksmart/rif-relay-contracts/dist/typechain-types/contracts/RelayHub';
+
+export type EIP712Domain = TypedMessage<EnvelopingMessageTypes>['domain'];
+
+export const eip712DomainType = [
+  { name: 'name', type: 'string' },
+  { name: 'version', type: 'string' },
+  { name: 'chainId', type: 'uint256' },
+  { name: 'verifyingContract', type: 'address' },
+];
+
+export const relayDataType = [
+  { name: 'gasPrice', type: 'uint256' },
+  { name: 'feesReceiver', type: 'address' },
+  { name: 'callForwarder', type: 'address' },
+  { name: 'callVerifier', type: 'address' },
+];
+
+export const requestType = [
+  { name: 'relayHub', type: 'address' },
+  { name: 'from', type: 'address' },
+  { name: 'to', type: 'address' },
+  { name: 'tokenContract', type: 'address' },
+  { name: 'value', type: 'uint256' },
+  { name: 'nonce', type: 'uint256' },
+  { name: 'tokenAmount', type: 'uint256' },
+  { name: 'tokenGas', type: 'uint256' },
+  { name: 'data', type: 'bytes' },
+  { name: 'relayData', type: 'RelayData' },
+];
+
+export const relayRequestType = [
+  ...relayDataType,
+  { name: 'gas', type: 'uint256' },
+];
+
+export const deployRequestType = [
+  ...relayDataType,
+  { name: 'recoverer', type: 'address' },
+  { name: 'index', type: 'uint256' },
+];
+
+export type EnvelopingMessageTypes = MessageTypes & {
+  RelayRequest: MessageTypeProperty[];
+  RelayData: MessageTypeProperty[];
+};
+
+export function getDomainSeparator(
+  verifyingContract: string,
+  chainId: number
+): EIP712Domain {
+  return {
+    name: 'string name,string version',
+    version: '2',
+    chainId: chainId,
+    verifyingContract: verifyingContract,
+  };
+}
+
+export const getDomainSeparatorHash = (
+  verifier: string,
+  chainId: number
+): string =>
+  TypedDataUtils.hashStruct(
+    'EIP712Domain',
+    getDomainSeparator(verifier, chainId),
+    { EIP712Domain: eip712DomainType },
+    SignTypedDataVersion.V4
+  ).toString('hex');
+
+type GetRequestDataFieldProps = {
+  chainId: number;
+  verifier: string;
+  relayRequest:
+    | EnvelopingTypes.RelayRequestStruct
+    | EnvelopingTypes.DeployRequestStruct;
+  requestTypes: MessageTypeProperty[];
+};
+
+export const getEnvelopingRequestDataV4Field = ({
+  chainId,
+  verifier, // is this separated from the relay request by design?
+  relayRequest,
+  requestTypes,
+}: GetRequestDataFieldProps): TypedMessage<EnvelopingMessageTypes> => ({
+  types: {
+    EIP712Domain: eip712DomainType,
+    RelayRequest: requestTypes,
+    RelayData: relayDataType,
+  },
+  primaryType: 'RelayRequest', // what if it's deploy request?
```

Comment:
Is this the correct primary type? Would have expected this to be `DeployRequest` for deploy requests...

Reply:
The name was standardized but we could change it. If we are going to change it, we need to change in the contracts too, in the factories to be specific.

Reply:
I mean, we have `RelayRequest` and `DeployRequest` types. Yet, here we hardcode it to the latter, and I just wonder: is that the expected primary type for all and if so, then why?

Reply:
Yes its expected to be the RelayRequest for the two types of transactions. Why Im not sure why they kept the same name, it could be changed, but that change includes a change in the contracts. 

This primary type that is declared here, serves to identify which is the struct has inside a nested struct. 

`RelayRequest(address relayHub,address from,address to,address tokenContract,address recoverer,uint256 value,uint256 nonce,uint256 tokenAmount,uint256 tokenGas,uint256 index,bytes data,RelayData relayData)RelayData(uint256 gasPrice,address feesReceiver,address callForwarder,address callVerifier)`

This line of code is inside the SmartWallet and SmartWalletFactory. As you can see the RelayRequest has inside of it the RelayData, and if we want we could change those names.